src/Microsoft.ML.Data/Transforms/HashTransform.cs

```diff
@@ -555,26 +581,56 @@ private ValueGetter<VBuffer<uint>> ComposeGetterVec(IRow input, int iinfo)
                     if (len > 0 && src.Length != len)
                         throw Host.Except("Hash transform expected {0} slots, but got {1}", len, src.Length);
 
+                    T[] values = src.Values;
+                    var srcIsDense = src.IsDense;
+                    // force-densify the input in case of float / double and ordered hash.
+                    if (!srcIsDense && ex.Ordered && (typeof(T) == typeof(float) || typeof(T) == typeof(double)))
+                    {
+                        if (denseValues == null)
+                            denseValues = new T[expectedSrcLength];
+                        values = denseValues;
+                        src.CopyTo(values);
+                        srcIsDense = true;
+                    }
+
                     var hashes = dst.Values;
-                    if (Utils.Size(hashes) < src.Count)
-                        hashes = new uint[src.Count];
+                    if (Utils.Size(hashes) < values.Length)
+                        hashes = new uint[values.Length];
 
-                    if (src.IsDense)
+                    if (srcIsDense)
                     {
-                        hasherDense(src.Count, null, src.Values, hashes, seed, mask);
+                        if (ex.Ordered)
```

Comment:
>ex.Ordered [](start = 28, length = 10)

So you've put a tremendous amount of logic that previously was handled *outside* the delegate, to inside the delegate. This cuts at the reason why we are composing delegates in the first place!

Consider, for example, `ex.Ordered`. Previously, if it was ordered, there was a single test *once* when the delegate was composed. Now this condition is checked every single time, despite the fact that for an instance of the cursor it will be constant!

Worse still are the presence of these type checks. Now on every single call, we are potentially checking where `typeof(T) == typeof(float)`.... even though that is definitely not going to change from call to call in the delegate!

Put all conditions you reasonably can put outside of the delegate, outside of the delegate. Just as a general rule. That's why we've structured the method the way we have.

Reply:
Consider reading this:

[https://github.com/dotnet/machinelearning/blob/master/docs/code/IDataViewImplementation.md#why-getgetter](https://github.com/dotnet/machinelearning/blob/master/docs/code/IDataViewImplementation.md#why-getgetter)

In particular the section about how we try to make things that *can* be verified only once, be verified only once, is important.

I feel like you're spending way, way too much effort trying to make this `ComposeGetterVecCore` work for all possible cases. You should probably just make a separate method for the float and double case (and other non-sparsity preserving cases) and be done with it.

The hash transform is important. We don't want perf degraded on it, certainly not the degree as was done here.

---
In reply to: [213158411](https://github.com/dotnet/machinelearning/pull/684#discussion_r213158411) [](ancestors = 213158411)

Reply:
Thank you Tom, great points, will re-do